Nama : Fadhail Athaillah Bima<br>
NIM : 24060121140172<br>
Nama Dataset : Nursery<br>
Link Dataset : https://archive.ics.uci.edu/ml/machine-learning-databases/spambase/spambase.data

# Import Libraries

In [1]:
# Import Libraries
import pandas as pd
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
import seaborn as sns
from sklearn.preprocessing import LabelEncoder, StandardScaler

# Memuat Dataset

In [2]:
# Memuat Dataset
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/spambase/spambase.data"

names = ['word_freq_cs', 'word_freq_meeting', 'word_freq_original', 'word_freq_project', 'word_freq_re', 'word_freq_edu', 'word_freq_table', 'word_freq_conference', 'class']

dataset = pd.read_csv(url, names=names)

## Preprocess the data

In [13]:
le = LabelEncoder()
for col in dataset.columns:
    dataset[col] = le.fit_transform(dataset[col])

# Dimensi Dataset

In [3]:
dataset.shape

(4601, 9)

jumlah baris dalam dataset adalah 4601 dan jumlah kolom dalam data set adalah 9

# Isi Dataset

In [4]:
dataset

word_freq_cs  \
0.00 0.64 0.64 0.0 0.32 0.00 0.00 0.00 0.00 0.00 0.00 0.64 0.00 0.00 0.00 0.32 0.00 1.29 1.93 0.00 0.96 0.0 0.00 0.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.00 0.0 0.0 0.00 0.00 0.00 0.00 0.0 0.0 0.000         0.000   
0.21 0.28 0.50 0.0 0.14 0.28 0.21 0.07 0.00 0.94 0.21 0.79 0.65 0.21 0.14 0.14 0.07 0.28 3.47 0.00 1.59 0.0 0.43 0.43 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.07 0.0 0.0 0.00 0.0 0.0 0.00 0.00 0.00 0.00 0.0 0.0 0.000         0.132   
0.06 0.00 0.71 0.0 1.23 0.19 0.19 0.12 0.64 0.25 0.38 0.45 0.12 0.00 1.75 0.06 0.06 1.03 1.36 0.32 0.51 0.0 1.16 0.06 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.06 0.0 0.0 0.12 0.00 0.06 0.06 0.0 0.0 0.010         0.143   
0.00 0.00 0.00 0.0 0.63 0.00 0.31 0.63 0.31 0.63 0.31 0.31 0.31 0.00 0.00 0.31 0.00 0.00 3.18 0.00 0.31 0.0 0.00 0.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.00 0.0 0.0 0.00 0.00 0.00 0.00 0.0 0.0 0.000         0.137   
                                                                                                                                                                                                                            0.000         0.135   
...                                                                                                                                                                                                                                         ...   
0.31 0.00 0.62 0.0 0.00 0.31 0.00 0.00 0.00 0.00 0.00 1.88 0.00 0.00 0.00 0.00 0.00 0.00 0.62 0.00 0.00 0.0 0.00 0.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.00 0.0 0.0 0.00 0.31 0.31 0.31 0.0 0.0 0.000         0.232   
0.00 0.00 0.00 0.0 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 6.00 0.00 2.00 0.0 0.00 0.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.00 0.0 0.0 0.00 0.00 0.00 2.00 0.0 0.0 0.000         0.000   
0.30 0.00 0.30 0.0 0.00 0.00 0.00 0.00 0.00 0.00 0.00 1.80 0.30 0.00 0.00 0.00 0.00 0.90 1.50 0.00 0.30 0.0 0.00 0.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.00 0.0 0.0 0.00 0.00 0.00 1.20 0.0 0.0 0.102         0.718   
0.96 0.00 0.00 0.0 0.32 0.00 0.00 0.00 0.00 0.00 0.00 0.32 0.00 0.00 0.00 0.00 0.00 0.00 1.93 0.00 0.32 0.0 0.00 0.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.00 0.0 0.0 0.00 0.32 0.00 0.32 0.0 0.0 0.000         0.057   
0.00 0.00 0.65 0.0 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.65 0.00 0.00 0.00 0.00 0.00 4.60 0.00 0.65 0.0 0.00 0.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.00 0.0 0.0 0.00 0.00 1.97 0.65 0.0 0.0 0.000         0.000   

                                                                                                                                                                                                                                   word_freq_meeting  \
0.00 0.64 0.64 0.0 0.32 0.00 0.00 0.00 0.00 0.00 0.00 0.64 0.00 0.00 0.00 0.32 0.00 1.29 1.93 0.00 0.96 0.0 0.00 0.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.00 0.0 0.0 0.00 0.00 0.00 0.00 0.0 0.0 0.000                0.0   
0.21 0.28 0.50 0.0 0.14 0.28 0.21 0.07 0.00 0.94 0.21 0.79 0.65 0.21 0.14 0.14 0.07 0.28 3.47 0.00 1.59 0.0 0.43 0.43 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.07 0.0 0.0 0.00 0.0 0.0 0.00 0.00 0.00 0.00 0.0 0.0 0.000                0.0   
0.06 0.00 0.71 0.0 1.23 0.19 0.19 0.12 0.64 0.25 0.38 0.45 0.12 0.00 1.75 0.06 0.06 1.03 1.36 0.32 0.51 0.0 1.16 0.06 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.06 0.0 0.0 0.12 0.00 0.06 0.06 0.0 0.0 0.010                0.0   
0.00 0.00 0.00 0.0 0.63 0.00 0.31 0.63 0.31 0.63 0.31 0.31 0.31 0.00 0.00 0.31 0.00 0.00 3.18 0.00 0.31 0.0 0.00 0.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.00 0.0 0.0 0.00 0.0 0.0 0.00 0.00 0.00 0.00 0.0 0.0 0.000                0.0   
                                                                      

Diatas merupakan tabel dari isi dataset

# Distribusi Kelas Data


In [ ]:
dataset.groupby('class').size()

class
0    4320
1    4266
2       2
3    4044
4     328
dtype: int64

### Pertemuan 2

## Import Model Selection

In [14]:
from sklearn import model_selection

##Train-Test Split

In [15]:
array = dataset.values

X = array[:,0:4]
Y = array[:,4]

validation_size = 0.20
seed = 7
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

X berisi fitur (features) dari dataset, yaitu kolom 0 hingga 4.<br>
Y berisi target (label) dari dataset, yaitu kolom 5.<br><br>
train_test_split adalah fungsi dari modul model_selection yang digunakan untuk membagi dataset menjadi data latih dan data uji.<br>
X_train dan Y_train adalah data latih untuk fitur dan target.<br>
X_validation dan Y_validation adalah data uji untuk fitur dan target.<br>
test_size=0.20 menunjukkan bahwa 20% dari dataset akan digunakan sebagai data uji.<br>
random_state digunakan untuk mengatur seed pada proses pembagian data agar hasilnya dapat direproduksi.

##  Menampilkan shape dari masing-masing subset data setelah dilakukan pembagian.

In [16]:
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (3680, 4)
X_validation : (921, 4)
Y_train : (3680,)
Y_validation : (921,)


## K-Folds Cross Validation

In [17]:
seed = 7

scoring = 'accuracy'

## Model Algoritma di Modul

In [18]:
#Import library algoritma
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB

#Spot Check Algorithms
models = []

models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))

## Evaluasi Algoritma Model

In [19]:
# evaluate each model in turn
results = []

names = []
for name, model in models:
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
  cv_results = model_selection.cross_val_score(model, X_train, Y_train,
  cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

KNN: 0.841848 (0.015362)
NB: 0.064130 (0.017577)
SVM: 0.837772 (0.013481)


Mengevaluasi kinerja beberapa model machine learning menggunakan validasi silang (cross-validation).<br>
Model yang dinilai:<br>


*   K-Nearest Neighbors (KNN): Akurasi rata-rata sekitar 30.79%, dengan deviasi standar sekitar 0.015362.
*   Naive Bayes (NB): Akurasi rata-rata sekitar 32.17%, dengan deviasi standar sekitar 0.017577.
*   Support Vector Machine (SVM): Akurasi rata-rata sekitar 28.74%, dengan deviasi standar sekitar 0.013481.

## Import Library untuk Testing

In [20]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

## Melakukan pengujian tentang keakuratan model Naive Bayes terhadap data

In [22]:
# Make predictions on validation dataset
KNN = KNeighborsClassifier()

KNN.fit(X_train, Y_train)

predictions = KNN.predict(X_validation)
print(accuracy_score(Y_validation, predictions))

print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.8295331161780674
[[757   0   0 ...   0   0   0]
 [  1   0   0 ...   0   0   0]
 [  1   0   0 ...   0   0   0]
 ...
 [  1   0   0 ...   0   0   0]
 [  1   0   0 ...   0   0   0]
 [  1   0   0 ...   0   0   0]]
              precision    recall  f1-score   support

           0       0.86      0.99      0.92       768
           1       0.00      0.00      0.00         1
           6       0.00      0.00      0.00         1
           7       0.00      0.00      0.00         1
           8       0.00      0.00      0.00         2
           9       0.00      0.00      0.00         2
          10       0.00      0.00      0.00         3
          12       0.00      0.00      0.00         4
          13       0.00      0.00      0.00         2
          14       1.00      1.00      1.00         1
          15       0.00      0.00      0.00         1
          16       0.00      0.00      0.00         1
          21       0.00      0.00      0.00         1
          22       0.00      0.0

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.

karena Naive Bayes memiliki nilai akurasi perkiraan terbesar, maka dilakukan pengujian mengenai keakuratan Model tersebut.<br>
Akurasi model pada dataset validasi adalah sekitar 30.67%.